### DISCUSSION TITLE
How to not duplicate user information in Database backend

### DISCUSSION THREAD

**Discussion by Enduriel** [maxcountryman/axum-login#211] [Q&A]
Let me start off by saying this crate looks awesome and it really seems to simplify the process of creating an auth system.

Unfortunately, I seem to be having some issues trying to use this system without creating some data duplication in my database, which I'm hoping to avoid, this is due to the way session data is stored using tower-sessions. I'm still very new to this so feel free to let me know if I'm misunderstanding something.

Basically, in the sqlite example, it seems like the User is stored directly in the session data. This means that if I have a separate table for users, data for users in that table would be duplicated in the session blob data. The obvious solution to this is to only store an ID or something like that in the session data which I can then use to get the full data later, but at that point I don't think the column even needs to be a BLOB, and should instead just become a reference to the id in the user table, unfortunately this doesn't really work with the more 'universal' data in the `Record`.

I'm trying to understand how to make this work in a way that makes sense in my head.

Maybe I can assume that the Record.data only contains the id to the user and only serialize that (seems bad), or store the ID as its own column and then BLOB the rest of the Data
Maybe there's another way I'm missing to go around this and have both a nice API and a logically sound DB backend for this setup

Would really appreciate some help.

Also wanted to say I really appreciate the examples being there, it's great to have them! The one bit of feedback I will give as someone looking at this stuff for the first time; the structure of the files was not intuitive for me and I would have appreciated a readme to understand why they are structured the way they are ^^.

**Comment by maxcountryman** [comment]
> Maybe there's another way I'm missing to go around this and have both a nice API and a logically sound DB backend for this setup

May I suggest that another option here is to not worry so much about what's stored in the session? The fact the session is persisted in the database is incidental: it could just as easily be in separate key-value store, in memory, etc.

Another way of saying this is that `axum-login` and `tower-sessions` aren't really intended as repositories of canonical entity data (the first is user session management and the second is session management itself).

> Basically, in the sqlite example, it seems like the User is stored directly in the session data.

While this is true, what's important is that `get_user` will hydrate the session on every request, so it's really only a cache of `User`. The next time a request is loaded, the value will be replaced with whatever your `get_user` implementation produces (which we assume is tied to some canonical representation; it's possible this is itself cached, e.g. to prevent overloading your database and so forth, but regardless it's going back to the logically sound definition, not providing a definition in and of itself).

**Reply by Enduriel** [reply]
This seems a little backwards to me. If I understand it correctly (and please do correct me if I am wrong) you're pretty much saying I shouldn't worry about what's stored in the session because it's thrown out anyway on a request, but at that point why is the user being stored in the session 'cache' at all?

Basically I'm wondering if it really makes sense to organize the type/data this way and is it necessary? Or is this just a workaround for the limitations of the existing tools.

**Reply by maxcountryman** [reply]
> because it's thrown out anyway on a request, but at that point why is the user being stored in the session 'cache' at all?

It's not thrown out, but overwritten. 

As for why store the user at all, the higher level idea is that your application will want to do something with the `AuthSession` during the course of processing a request and building a response (this could be anything from gate access to resources to displaying things like a username in your UI).

> Basically I'm wondering if it really makes sense to organize the type/data this way

It does. 🙂

**Reply by Enduriel** [reply]
So I ended up looking at the code, and it seems like I completely misunderstood how it worked and we misunderstood each other. I assumed that each session stored the entire `Backend::User` struct, rather than just the `Id` and `auth_hash`. All clear now and it makes perfect sense, sorry for the confusion.

**Reply by maxcountryman** [reply]
I'm a little bit concerned future folks might see this and draw the wrong conclusions, because you seem to be confused, so I'll restate what I said before: [`AuthSession` _does_ provide the `User`](https://docs.rs/axum-login/latest/axum_login/struct.AuthSession.html#structfield.user). But as I already explained, you should think of this as a cached value, not a canonical representation.

**Reply by Enduriel** [reply]
Let me try and clarify again, my only concern was about how this data was stored in the backend. Specifically the fact that I assumed the User struct was serialized into `tower_sessions` `Session::data` completely, rather than just its Id (and auth_hash).

From what I understand, you are talking about how it works once the data has been retrieved from the backend and you're working with a particular request. I don't fully understand how it works right now to be fair, but it's not relevant to my original question.

ETA: Reading [this](https://github.com/maxcountryman/axum-login/blob/e9597526afbe6325826b24c04af645f1e52eeae6/axum-login/src/session.rs#L110) is what cleared it up for me
